### ISSUE TITLE
apex infra does not pass the function name with -var

### ISSUE THREAD

**Issue by rantav** [apex/apex#335]
According to documentation, running `apex infra plan` or `apex infra apply` should pass a variable to terraform with the ARN of the functions, for example `${var.apex_function_xxx}` should be defined in TF. 

According to my tests - it isn't. 

Here's a repo with a simple repro: https://github.com/rantav/apex-bug-repro-335

When running apex infra plan you'd notice that TF asks for a value for the var `apex_function_hello`. To my understanding, this shouldn't happen if apex had passed the value for the var. 

Here are two examples. The first demonstrates the issue (TF asks for my input)

```
$ apex infra plan
var.apex_function_hello
  Enter a value: xxx       <<<---- TF asks for my input

Refreshing Terraform state prior to plan...


No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

The second shows that using env variables, TF shuts up....
```
$ TF_VAR_apex_function_hello=xxx apex infra plan
Refreshing Terraform state prior to plan...


No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

My hope is that - I'm not not using it correctly. Did I get the correct var name? Is there something wrong with my TF code? Or is this a bug?

**Comment by mthenw** [comment]
@rantav is it still an issue? I guess the problem was that you didn't defined empty variable in TF(?).

**Comment by rantav** [comment]
Yes it is still an issue. 
I did add an empty variable in TF, see https://github.com/rantav/apex-bug-repro-335/blob/master/infrastructure/main.tf#L5

Did I do this incorrectly?

**Comment by mthenw** [comment]
Looks fine, could you run `apex infra plan --log-level debug` and paste output here.

**Comment by tj** [comment]
hmm that's kind of annoying if TF makes you use `variable {}` for that. I have my fn ARNs hard-coded in ATM

**Comment by rantav** [comment]
Running apex infra plan (or apply) requires providing the value for the variable `apex_function_hello`:


```
$ apex infra plan --log-level debug
   • couldn't read role from infrastructure: exit status 1
   • loading functions in functions
   • loading function in functions/hello
   • inferring runtime         function=hello
   • inferred runtime          function=hello runtime=nodejs
   • terraform                 args=[plan]
var.apex_function_hello
  Enter a value:
```
(I stopped here)

**Comment by rantav** [comment]
And BTW:
```
$ terraform -v
Terraform v0.6.12
```


and 
```
$ apex version
Apex version 0.7.0
```

**Comment by mthenw** [comment]
Just to make sure, have you deployed functions before using functions' ARNs in TF configs?

**Comment by rantav** [comment]
Hi @mthenw yes, the function had been deployed. 
Look, I'm sure it's just a simple oversight from my end, but beats me, I can't figure out what I'm doing wrong... Does the `-var` work for you? 

Also, if seems that the output of `--log-level debug` includes the line `• terraform                 args=[plan]` so does that mean that the `-var` is indeed not passed?

**Comment by mthenw** [comment]
OK I figured out what is wrong. Based on your repro repo there is no role passed. If there is no role apex doesn't load functions. Not sure how you created `infrastructure` folder but that wasn't generated with `apex init`. `apex init` inits infrastructure repo with TF module responsible for creating IAM role that is read by apex to deploy functions.

Basically the flow should look like this:

- `apex init` 
- `apex infra apply`
- `apex deploy`
- some changes to TF configs using functions arns
- `apex infra apply`
- ...

For sure we need to work on better docs about that.

**Comment by rantav** [comment]
Hmm, I apologize, I did add a role later in development, but I did not push the commit. I did push it just now though.

However, even after having added the role (I can't deploy without a role, BTW) - I still can't get the ARNs passed to TF. Did you? 

BTW - the flow is a bit perplexed. As someone that now understands the implemantation details I can see why you need `init` and then `infra apply` and then `deploy` and then some changes and then again `infra apply` however, as a noob - this still seems somewhat confusing, usability wise...

Anyway, bottom line for me - this still isn't working. Does it work for you? (as mentioned, I added the role but this role only makes sense under my AWS account not yours ;))

**Comment by rantav** [comment]
OK, I found the problem, it was my bad, thanks @mthenw for your help!

One has to remember a few things: 

1. You have to deploy the function before it's ARN can be used. (ARNs are created only after a resource actually exists in AWS)
2. One of the errors that stumbled me was that at some point I did not provide the region, so even though apex was going to fail on that if I was going to run `apex deploy` without a region - running `apex infra` does not fail, but it also does not pass the -var ARN as expected. Fixing this was easy - just add the aws region, for example `AWS_REGION=us-west-2 apex infra plan`

My suggestion would be though @mthenw to check for missing region (and other required params such as aws key) before calling to TF, similar to the way you check when running `apex deploy`. I'm happy to open a bug to track this issue if you agree. 

I'm going to go ahead and close this issue right now.
